server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java

```diff
@@ -53,6 +54,7 @@
     private final BigArrays bigArrays;
     private final IntConsumer multiBucketConsumer;
     private IntArray docCounts;
+    protected final DocCountProvider docCountProvider;
```

Comment:
We could use a `NumericDocValues` directly and have the logic in a protected function like `getDocCountForDoc(int)` ? We don't really need the new interface and classes ?

Reply:
Method `DocCountProvider#getDocCount(int)` is used by `BucketsAggregator` and `SortedDocsProducer` classes

https://github.com/elastic/elasticsearch/blob/7b7ca43d6aa9bac621f3977c519a407122e10623/server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java#L98

https://github.com/elastic/elasticsearch/blob/7b7ca43d6aa9bac621f3977c519a407122e10623/server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/SortedDocsProducer.java#L72

I think the easier way to share the code of reading `_doc_count` field is to encapsulate it in a small simple class and call it from both `BucketsAggregator` and `SortedDocsProducer` classes. To simplify the class even more, I have modified `DocCountProvider` interface and made it a concrete class with the implementation.